README.md
# Flutter Console Coverage Test

This small dart tools is used to generate Flutter Coverage Test report to console

## How to install
Add a line like this to your package's pubspec.yaml (and run an implicit flutter pub get):
```
dev_dependencies:
  test_cov_console: ^0.0.4
```

## How to run
### run the following command to make sure all flutter library is up-to-date
```
flutter pub get
Running "flutter pub get" in coverage...                            0.5s
```
### run the following command to generate lcov.info on coverage directory
```
flutter test --coverage
00:02 +1: All tests passed!
```
### run the tool to generate report from lcov.info
```
flutter pub run test_cov_console
---------------------------------------------|---------|---------|---------|-------------------|
File                                         |% Branch | % Funcs | % Lines | Uncovered Line #s |
---------------------------------------------|---------|---------|---------|-------------------|
lib/src/                                     |         |         |         |                   |
 print_coverage.dart                         |  100.00 |  100.00 |   80.87 |...,206,207,208,209|
lib/                                         |         |         |         |                   |
 test_cov_console.dart                       |    0.00 |    0.00 |    0.00 |    no unit testing|
---------------------------------------------|---------|---------|---------|-------------------|
 All files with unit testing                 |  100.00 |  100.00 |   80.87 |                   |
---------------------------------------------|---------|---------|---------|-------------------|
```


CHANGELOG.md
## 0.0.4 - 16 May 2021

* Print all dart files, not only that reported on lcov.info (has unit test file)
* Print total coverage for all files with unit testing.

## 0.0.3 - 15 May 2021

* Add example project
* Add simple documentation

## 0.0.2 - 14 May 2021

* Fixed typo issue.

## 0.0.1 - 14 May 2021

* Initial code


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:coverage/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/test_cov_console_test.dart
import 'dart:async';

import 'package:flutter_test/flutter_test.dart';
import 'package:test_cov_console/test_cov_console.dart';

var log = <String>[];

void main() {
  final lines = lcovFile.split('\n');
  final printout = output.split('\n');
  group('printCoverage', () {
    test('should print out with correct format & values', overridePrint(() {
      printCoverage(lines, files);
      expect(log, printout);
    }));
  });
}

void Function() overridePrint(void testFn()) => () {
      final spec = ZoneSpecification(print: (_, __, ___, String msg) {
        // Add to log instead of printing to stdout
        log.add(msg);
      });
      return Zone.current.fork(specification: spec).run<void>(testFn);
    };

const List<String> files = [
  'lib/src/print_coverage.dart',
  'lib/test_cov_console.dart',
];
const String lcovFile = '''
SF:lib/src/print_coverage.dart
DA:8,1
DA:9,1
DA:10,1
DA:12,1
DA:13,3
DA:14,1
DA:15,1
DA:16,1
DA:17,1
DA:18,1
DA:19,1
DA:20,1
DA:21,1
DA:22,1
DA:23,1
DA:24,1
DA:25,1
DA:26,1
DA:27,1
DA:28,3
DA:29,1
DA:30,1
DA:32,1
DA:33,1
DA:36,1
DA:37,1
DA:38,3
DA:39,3
DA:42,1
DA:43,2
DA:45,1
DA:46,2
DA:48,1
DA:49,0
DA:51,1
DA:52,0
DA:54,1
DA:55,0
DA:57,1
DA:58,0
DA:60,1
DA:61,0
DA:62,0
DA:63,0
DA:66,1
DA:68,1
DA:69,1
DA:70,1
DA:71,2
DA:72,2
DA:73,0
DA:78,1
DA:79,1
DA:80,2
DA:81,1
DA:94,1
DA:107,1
DA:110,1
DA:111,1
DA:114,4
DA:117,1
DA:118,2
DA:120,0
DA:123,1
DA:125,3
DA:126,1
DA:127,1
DA:128,1
DA:129,1
LF:69
LH:60
end_of_record
''';

const output = '''
---------------------------------------------|---------|---------|---------|-------------------|
File                                         |% Branch | % Funcs | % Lines | Uncovered Line #s |
---------------------------------------------|---------|---------|---------|-------------------|
lib/src/                                     |         |         |         |                   |
 print_coverage.dart                         |  100.00 |  100.00 |   86.96 |...,61,62,63,73,120|
lib/                                         |         |         |         |                   |
 test_cov_console.dart                       |    0.00 |    0.00 |    0.00 |    no unit testing|
---------------------------------------------|---------|---------|---------|-------------------|
 All files with unit testing                 |  100.00 |  100.00 |   86.96 |                   |
---------------------------------------------|---------|---------|---------|-------------------|''';


lib/coverage.dart
import 'dart:io';

const fileLen = 45;
const percentLen = 9;
const uncoverLen = 19;

Future<void> main(List<String> args) async {
  if (args.length == 0) {
    return;
  }
  final lcovFile = args[0];
  final lines = await File(lcovFile).readAsLines();

  _print('-', '-', '-', '-', '-', '-');
  _print(
      'File', '% Branch ', '% Funcs ', '% Lines ', 'Uncovered Line #s ', ' ');
  _print('-', '-', '-', '-', '-', '-');
  lines.fold([0, 0, 0, 0, 0, 0, '', '', '', ''], (List<dynamic> data, line) {
    var functionFound = data[0];
    var functionHit = data[1];
    var linesFound = data[2];
    var linesHit = data[3];
    var branchFound = data[4];
    var branchHit = data[5];
    String uncoveredLines = data[6];
    String uncoveredBranch = data[7];
    String fileName = data[8];
    String directory = data[9];
    final values = line.split(':');
    switch (values[0]) {
      case 'SF':
        final fullFileName = values.last;
        fileName = fullFileName.replaceAll('/', '\\').split('\\').last;
        final dir = fullFileName.replaceAll(fileName, '');
        if (dir != directory) {
          directory = dir;
          _print(_formatString(directory, fileLen, ''), ' ', ' ', ' ', ' ', ' ');
        }
        break;
      case 'DA':
        if (line.endsWith('0')) {
          uncoveredLines = (uncoveredLines != '' ? '$uncoveredLines,' : '') +
              values[1].split(',')[0];
        }
        break;
      case 'LF':
        linesFound = int.parse(values[1]);
        break;
      case 'LH':
        linesHit = int.parse(values[1]);
        break;
      case 'FNF':
        functionFound = int.parse(values[1]);
        break;
      case 'FNH':
        functionHit = int.parse(values[1]);
        break;
      case 'BRF':
        branchFound = int.parse(values[1]);
        break;
      case 'BRH':
        branchHit = int.parse(values[1]);
        break;
      case 'BRDA':
        if (line.endsWith('0')) {
          uncoveredBranch = (uncoveredBranch != '' ? '$uncoveredBranch,' : '') +
              values[1].split(',')[0];
        }
        break;
      case 'end_of_record':
        {
          final functions = _formatPercent(functionHit, functionFound);
          final lines = _formatPercent(linesHit, linesFound);
          final branch = _formatPercent(branchHit, branchFound);
          if (functions.trim() == '100.00' &&
              lines.trim() == '100.00' &&
              branch.trim() == '100.00') {
            uncoveredLines = '';
            uncoveredBranch = '';
          }
          var uncovered =
          uncoveredLines.isEmpty ? uncoveredBranch : uncoveredLines;
          uncovered = _formatString(uncovered, uncoverLen, '...');
          final file = _formatString(' $fileName', fileLen, '');
          _print(file, branch, functions, lines, uncovered, ' ');
          linesFound = 0;
          linesHit = 0;
          functionHit = 0;
          functionFound = 0;
          branchHit = 0;
          branchFound = 0;
          uncoveredLines = '';
          uncoveredBranch = '';
        }
        break;
    }

    return [
      functionFound,
      functionHit,
      linesFound,
      linesHit,
      branchFound,
      branchHit,
      uncoveredLines,
      uncoveredBranch,
      fileName,
      directory,
    ];
  });
  _print('-', '-', '-', '-', '-', '-');
}

String _formatPercent(hit, found) {
  if (found == 0) {
    return '100.00 ';
  }
  return '${(hit / found * 100).toStringAsFixed(2)} ';
}

String _formatString(String input, int length, String more) {
  return input.length <= length
      ? input
      : '$more${input.substring(input.length - length + more.length)}';
}

void _print(file, branch, function, lines, uncovered, filler) {
  print('${file.padRight(fileLen, filler)}|'
      '${branch.padLeft(percentLen, filler)}|'
      '${function.padLeft(percentLen, filler)}|'
      '${lines.padLeft(percentLen, filler)}|'
      '${uncovered.padLeft(uncoverLen, filler)}|');
}


lib/src/print_coverage.dart
// Copyright (c) 2021, I Made Mudita. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

import 'dart:io';

const _fileLen = 45;
const _percentLen = 9;
const _uncoverLen = 19;
const _hundred = '100.00';
const _zero = '0.00 ';
const _dart = 'dart';
const _slash = '/';
const _bSlash = '\\';

class FileEntity {
  String fileName = '';
  String directory = '';

  FileEntity(String fullName) {
    fileName = fullName.split(_slash).last;
    directory = fullName.replaceAll(fileName, '');
  }

  int compareTo(FileEntity other) {
    if (directory.compareTo(other.directory) < 0) {
      return -1;
    }
    if (directory.compareTo(other.directory) > 0) {
      return 1;
    }
    return fileName.compareTo(other.fileName);
  }
}

class _Data {
  int functionFound = 0;
  int functionHit = 0;
  int linesFound = 0;
  int linesHit = 0;
  int branchFound = 0;
  int branchHit = 0;
  String uncoveredLines = '';
  String uncoveredBranch = '';
  FileEntity file;

  _Data(FileEntity file) {
    this.file = file;
  }

  String getFileName() => file.fileName;

  String getDirectory() => file.directory;

  void total(_Data data) {
    functionFound += data.functionFound;
    functionHit += data.functionHit;
    linesFound += data.linesFound;
    linesHit += data.linesHit;
    branchFound += data.branchFound;
    branchHit += data.branchHit;
  }
}

void printCoverage(List<String> lines, List<FileEntity> files) {
  var idx = 0;
  _print('-', '-', '-', '-', '-', '-');
  _print(
      'File', '% Branch ', '% Funcs ', '% Lines ', 'Uncovered Line #s ', ' ');
  _print('-', '-', '-', '-', '-', '-');
  final result = lines
      .fold(<_Data>[_Data(FileEntity('')), _Data(FileEntity(''))],
          (List<_Data> data, line) {
    var data0 = data[0];
    final values = line.split(':');
    switch (values[0]) {
      case 'SF':
        final file = FileEntity(values.last.replaceAll(_bSlash, _slash));
        for (var i = idx; i < files.length; i++) {
          idx = i;
          if (file.compareTo(files[i]) > 0) {
            _printDir(files[i], data0.getDirectory(), true);
            data0.file.directory = files[i].directory;
          } else {
            break;
          }
        }
        if ((idx < files.length && file.compareTo(files[idx]) == 0) ||
            (idx == (files.length - 1) && file.compareTo(files[idx]) > 0)) {
          idx = idx + 1;
        }
        final result = _printDir(file, data0.getDirectory(), false);
        data0.file = result;
        break;
      case 'DA':
        if (line.endsWith('0')) {
          data0.uncoveredLines =
              (data0.uncoveredLines != '' ? '${data0.uncoveredLines},' : '') +
                  values[1].split(',')[0];
        }
        break;
      case 'LF':
        data0.linesFound = int.parse(values[1]);
        break;
      case 'LH':
        data0.linesHit = int.parse(values[1]);
        break;
      case 'FNF':
        data0.functionFound = int.parse(values[1]);
        break;
      case 'FNH':
        data0.functionHit = int.parse(values[1]);
        break;
      case 'BRF':
        data0.branchFound = int.parse(values[1]);
        break;
      case 'BRH':
        data0.branchHit = int.parse(values[1]);
        break;
      case 'BRDA':
        if (line.endsWith('0')) {
          data0.uncoveredBranch =
              (data0.uncoveredBranch != '' ? '${data0.uncoveredBranch},' : '') +
                  values[1].split(',')[0];
        }
        break;
      case 'end_of_record':
        {
          data0 = _printFile(data0);
          data[1].total(data0);
          data0 = _Data(data0.file);
        }
        break;
    }

    return [data0, data[1]];
  });
  if (idx < files.length) {
    for (var i = idx; i < files.length; i++) {
      _printDir(files[i], result[0].getDirectory(), true);
    }
  }
  _print('-', '-', '-', '-', '-', '-');
  result[1].file = FileEntity('All files with unit testing');
  _printFile(result[1]);
  _print('-', '-', '-', '-', '-', '-');
}

FileEntity _printDir(FileEntity file, String directory, bool printFile) {
  if (file.directory != directory) {
    _print(
        _formatString(file.directory, _fileLen, ''), ' ', ' ', ' ', ' ', ' ');
  }
  if (printFile) {
    _print(' ${file.fileName}', _zero, _zero, _zero, 'no unit testing', ' ');
  }
  return file;
}

_Data _printFile(_Data data0) {
  final functions = _formatPercent(data0.functionHit, data0.functionFound);
  final lines = _formatPercent(data0.linesHit, data0.linesFound);
  final branch = _formatPercent(data0.branchHit, data0.branchFound);
  if (functions.trim() == _hundred &&
      lines.trim() == _hundred &&
      branch.trim() == _hundred) {
    data0.uncoveredLines = '';
    data0.uncoveredBranch = '';
  }
  var uncovered = data0.uncoveredLines.isEmpty
      ? data0.uncoveredBranch
      : data0.uncoveredLines;
  uncovered = _formatString(uncovered, _uncoverLen, '...');
  final file = _formatString(' ${data0.getFileName()}', _fileLen, '');
  _print(file, branch, functions, lines, uncovered, ' ');

  return data0;
}

String _formatPercent(int hit, int found) {
  if (found == 0) {
    return '$_hundred ';
  }
  return '${(hit / found * 100).toStringAsFixed(2)} ';
}

String _formatString(String input, int length, String more) {
  return input.length <= length
      ? input
      : '$more${input.substring(input.length - length + more.length)}';
}

void _print(String file, String branch, String function, String lines,
    String uncovered, String filler) {
  print('${file.padRight(_fileLen, filler)}|'
      '${branch.padLeft(_percentLen, filler)}|'
      '${function.padLeft(_percentLen, filler)}|'
      '${lines.padLeft(_percentLen, filler)}|'
      '${uncovered.padLeft(_uncoverLen, filler)}|');
}

Future<List<FileEntity>> getFiles(String path) async {
  final dir = Directory(path);
  final files = await dir.list(recursive: true).toList();
  final List<FileEntity> list = [];
  files.forEach((element) {
    final String file = element.uri.toString();
    if (file.split('.').last == _dart) {
      final file = FileEntity(element.uri.toString());
      list.add(file);
    }
  });
  return list;
}
